### ISSUE TITLE
 env times, porta times and freqs should be log scale

### ISSUE THREAD

**Issue by magnetophon** [magnetophon/DigiDrie#11]

**Comment by magnetophon** [comment]
up/down time and PMs can be log

**Comment by ryukau** [comment]
From the chat:

```
23:58
bart
can we make that a special metadata that doesn't exist in faust.  those get ignored by the compiler, so if someone needs to use the raw faust code, they only get log on the parameters that atually need to be log.  All the other params get the new metadata, so will behave linear in the raw faust, without your gui
So for example: filter Q gets [scale:ryukau], attack gets [scale:log], and most params get no scale
```

Something like this is also possible:

```faust
value = hslider("name[scale:log][digidrie_scale:some_scale]", /* ... */);
```

With `faust -json`, this generates following json:

```json
  "meta": [
    { "digidrie_scale": "some_scale" },
    { "scale": "log" }
  ]
```

**Comment by ryukau** [comment]
Following mappings are available in my UI.

- `int`
- `linear`
- `spoly`
- `elliptic`
- `log`
- `decibel`

`int` can be used for option menu, like `type` parameter in DigiDrie. Also adding `[scale:int]` to `button` and `checkbox` helps a lot.

`spoly` and `elliptic` are both S-curve. There's subtle difference but likely negligible.

`log` is internally using `pow`. So it also covers Faust's `exp` case, if I understand correctly.

`decibel` is a bit tricky. This maps [0, 1] to decibel scale, then maps decibel value to amplitude. For example, if I set a `decibel` scale in range of [-30, 0] dB,

```
- 0   maps to -30 dB, then outputs amplitude of 0.0316.
- 0.5 maps to -15 dB, them outputs amplitude of 0.1778.
- 1   maps to   0 dB, then outputs amplitude of 1.
```

In other words, if you add `decibel` as scale metadata, expect to receive amplitude.

**Comment by magnetophon** [comment]
I added [scale:log] to these parameters in 01c7c9e172e642b26a37f3dd9c74ec05bb18b0ee.

Which other parameters should I change?

**Comment by ryukau** [comment]
Thank you!

For other parameters, it probably requires some experimentation. So we can come back to this issue after I made a working UI.

I have a question. Does it breaks Faust code by adding non-standard scale metadata? I'm considering to add `[scale:int]` to parameters using `button`, `checkbox` and `*slider` which works as option menu. This helps to automate code generation.

**Comment by magnetophon** [comment]
> Does it breaks Faust code by adding non-standard scale metadata?

No, that's fine.

There is already a menu in faust though.
Look at [type](https://github.com/magnetophon/DigiDrie/blob/01c7c9e172e642b26a37f3dd9c74ec05bb18b0ee/lib/GUI.lib#L97) for example.
`faust2jack` doesn't support this, but `faust2jaqt` does.

**Comment by ryukau** [comment]
I was thinking exactly about type. I'd like to have:

```faust
type = vslider("type[scale:int][style:menu{/* ... */}] /* ... */);
```

This makes querying scales just becomes `element["meta"]["scale"]` in python. Otherwise I have to add branch for `menu`.

If you are OK with this, I'll add these changes to `master` branch.

**Comment by magnetophon** [comment]
I've been making lots of other changes, so it's easier if I add it.

**Comment by ryukau** [comment]
OK. I'll leave it to you.

**Comment by magnetophon** [comment]
Done.

I also added [scale: log] to the PM params as suggested in the chat.

Let me know if there's anything I forgot regarding this issue.
